libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,567 @@
+				// it further up once more.
+				// This would mean not stopping at _block == target above or even doing visited.clear() here, revisiting the entire graph.
+				// This is a tradeoff between the runtime of this process and the optimality of the result.
+				// Also note that while visiting the entire graph again *can* be helpful, it can also be detrimental.
+				// Also note that for some reason using visited.clear() is incompatible with skipping the revisit
+				// of already seen exit layouts above, I'm not sure yet why.
+			}
+	}
+
+	stitchConditionalJumps(_entry);
+	fixStackTooDeep(_entry);
+}
+
+Stack StackLayoutGenerator::combineStack(Stack const& _stack1, Stack const& _stack2)
+{
+	if (_stack1.empty())
+		return _stack2;
+	if (_stack2.empty())
+		return _stack1;
+
+	// TODO: there is probably a better way than brute-forcing. This has n! complexity or worse, so
+	// we can't keep it like this.
+
+	Stack commonPrefix;
+	for (auto&& [slot1, slot2]: ranges::zip_view(_stack1, _stack2))
+	{
+		if (!(slot1 == slot2))
+			break;
+		commonPrefix.emplace_back(slot1);
+	}
+	Stack stack1Tail = _stack1 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+	Stack stack2Tail = _stack2 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+
+	Stack candidate;
+	for (auto slot: stack1Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	for (auto slot: stack2Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	cxx20::erase_if(candidate, [](StackSlot const& slot) {
+		return holds_alternative<LiteralSlot>(slot) || holds_alternative<FunctionCallReturnLabelSlot>(slot);
+	});
+
+	std::map<size_t, Stack> sortedCandidates;
+
+	// TODO: surprisingly this works for rather comparably large candidate size, but we should probably
+	// set up some limit, since this will quickly explode otherwise.
+	// Ideally we would then have a better fallback mechanism - although returning any naive union of both stacks
+	// like ``candidate`` itself may just be fine.
+	//	if (candidate.size() > 8)
+	//		return candidate;
+
+	auto evaluate = [&](Stack const& _candidate) -> size_t {
+		size_t numOps = 0;
+		Stack testStack = _candidate;
+		auto swap = [&](unsigned _swapDepth) { ++numOps; if (_swapDepth > 16) numOps += 1000; };
+		auto dup = [&](unsigned _dupDepth) { ++numOps; if (_dupDepth > 16) numOps += 1000; };
+		auto push = [&](StackSlot const& _slot) {
+			if (!canBeFreelyGenerated(_slot))
+			{
+				auto offsetInPrefix = util::findOffset(commonPrefix, _slot);
+				yulAssert(offsetInPrefix, "");
+				// Effectively this is a dup.
+				++numOps;
+				// TODO: Verify that this is correct. The idea is to penalize dupping stuff up that's too deep in
+				//       the prefix at this point.
+				if (commonPrefix.size() + testStack.size() - *offsetInPrefix > 16)
+					numOps += 1000;
+			}
+		};
+		createStackLayout(testStack, stack1Tail, swap, dup, push, [&](){} );
```

Comment:
```suggestion
		createStackLayout(testStack, stack1Tail, swap, dup, push, pop);
```
Naming it would make it more clear. Also the `&` can be skipped.

Also, why are we not counting for `pop`?

Reply:
The way the algorithm works, you'll always end up with the same amount of ``pop``s, no matter which permutation you start from - and it'd be a weird algorithm if that wasn't the case - it'd need to `dup` or `push` some garbage only to pop it again :-).
The same actually holds for pushing and dupping, we only really need to count ``swap``s... on the other hand we might as well also count ``pop``s, since it won't change the overall result.